File: DSA/Guaranteed Placement Course_C++/quickSort.cpp
#include <iostream>
using namespace std;

int partition(int arr[], int start, int end)
{
    int pivot = arr[end];
    int i = start - 1;
    for (int j = start; j < end; j++)
    {
        if (arr[j] < pivot)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[end]);
    return i + 1;
}

void quickSort(int arr[], int start, int end)
{
    if (start < end)
    {
        int pivot = partition(arr, start, end);
        quickSort(arr, start, pivot - 1);
        quickSort(arr, pivot + 1, end);
    }
}

int main()
{
    int arr[] = {1, 6, 3, 7, 2, 8, 9, 5, 4};
    quickSort(arr, 0, 8);
    for (int i = 0; i < 9; i++)
    {
        cout << arr[i] << " ";
    }

    return 0;
}

File: JAVA/Test/Test/SR/quickSort.java
package SR;

public class quickSort {
    public static int partition(int[] arr, int start, int end) {
        int pivot = arr[end];
        int i = start - 1;
        ;
        for (int j = start; j < end; j++) {
            if (arr[j] < pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        i++;
        int temp = arr[i];
        arr[i] = arr[end];
        arr[end] = temp;
        return i;
    }

    public static void QuickSort(int[] arr, int start, int end) {
        if (start < end) {
            int pivot = partition(arr, start, end);
            QuickSort(arr, start, pivot - 1);
            QuickSort(arr, pivot + 1, end);
        }
    }

    public static void main(String[] args) {

        int[] arr = { 2, 1, 5, 6, 4, 3 };
        int n = arr.length;
        QuickSort(arr, 0, n - 1);
        for (int i = 0; i < n; i++)
            System.out.print(arr[i] + " ");
    }
}
